design/metadata_kv/index.md

```diff
@@ -0,0 +1,187 @@
+
+This proposal describes an alternative implementation for the **mutable** metadata.
+Currently, this type of metadata is stored on (and relies on the guarantees provided by) a PostgreSQL database.
+
+In this proposal, we detail a relatively narrow database abstraction - one that could be satisfied by a wide variety of database systems.
+We also detail how lakeFS operates on top of this abstraction, instead of relying on PostgreSQL specific behavior
+
+## Goals
+
+1. Make it easy to run lakeFS as a managed service in a multitenant environment
+1. Allow running production lakeFS systems on top of a database that operations teams are capable of managing, as not all ops teams are versed in scaling PostgreSQL.
+1. Increase users trust in lakeFS, in terms of positioning: PostgreSQL is apparently not the first DB that comes to mind in relation to scalability
+1. Make lakeFS easier to experiment with: Allow running a local lakeFS server without any external dependencies (e.g. no more `docker-compose` in the quickstart)
+
+## Non-Goals
+
+1. Improve performance, latency or throughput of the system
+1. Provide new features, capabilities or guarantees that weren't previously possible in the lakeFS API/UI/CLI
+
+## Design
+
+At the heart of the design, is a simple Key/Value interface, used for all mutable metadata management.
+
+### Key/Value Store interface
+
+This is roughly the API:
+
+```go
+type Store interface {
+    // Get returns a value for the given key, or ErrNotFound if key doesn't exist
+    Get(key []byte) (value []byte, err error)
+    // Scan returns an iterator that scans keys in byte order, starting at or after the `start` position
+    Scan(start []byte) (iter KeyValueIterator, err error)
+    // Set stores the given value, overwriting an existing value if one exists
+    Set(key, value []byte) error
+    // Delete will delete the key/value at key, if any
+    Delete(key []byte) error
+    // SetIf returns an ErrPredicateFailed error if the valuePredicate passed
+    //  doesn't match the currently stored value.
+    SetIf(key, value, valuePredicate []byte) error
+}
+```
+
+Note: This API is roughly the one needed and is subject to change/tweaking. It is meant to illustrate the required capabilities in order to build a functioning lakeFS system on top.
+
+Writing a complete implementation of this API with the required guarantees is possible for a myriad of different database engines, such as:
+
+- PostgreSQL
+- MySQL
+- Embedded Pebble/RocksDB (great option for a "quickstart" environment?)
+- MongoDB
+- AWS DynamoDB
+- FoundationDB
+- Azure Cosmos
+- Azure Blob Store
+- Google BigTable
+- Google Spanner
+- Google Cloud Storage
+- HBase
+- Cassandra (or compatible systems such as ScyllaDB)
+- Raft (embedded, or server implementations such as Consul/ETCd)
+- Persistent Redis (Redis Enterprise, AWS MemoryDB)
+- Simple in-memory tree (for simplifying and speeding up tests?)
```

Comment:
Can we include a list (or multiple lists) of required primitives for the KV store?

Reply:
Added! thanks.